src/contract/deploy.rs

```diff
@@ -102,6 +103,50 @@ impl<T: Transport> Builder<T> {
             waiting,
         })
     }
+
+    /// Execute deployment passing code and contructor parameters.
+    pub fn execute<P, V>(self, code: V, params: P, from: Address) -> Result<PendingContract<T>, ethabi::Error>
+    where
+        P: Tokenize,
+        V: AsRef<str>,
+        T: Transport,
+    {
+        let f = |tx, eth: T, poll_interval, confirmations| {
+            confirm::send_transaction_with_confirmation(eth, tx, poll_interval, confirmations)
+        };
+        self.execute_core(code, params, from, f)
+    }
+
+    /// Execute deployment passing code and contructor parameters.
+    ///
+    /// Unlike the above `execute`, this method uses
+    /// `sign_raw_transaction_with_confirmation` instead of
+    /// `sign_transaction_with_confirmation`, which requires the account from
+    /// which the transaction is sent to be unlocked.
+    pub fn execute_no_unlock<P, V>(
+        self,
+        code: V,
+        params: P,
+        from: Address,
+        password: &str,
+        web3: crate::Web3<T>,
+    ) -> Result<PendingContract<T>, ethabi::Error>
+    where
+        P: Tokenize,
+        V: AsRef<str>,
+        T: Transport,
+    {
+        let f = |tx, eth: T, poll_interval, confirmations| {
+            let raw_tx = web3
+                .personal()
+                .sign_transaction(tx, password)
+                .wait()
```

Comment:
That is making a synchronous call, it's not acceptable in async context.

Reply:
Then how should I fetch the value that `web3.personal().sign_transaction()` returns?

I can chain this computation with `and_then(move |signed_tx| {send_raw_transaction_with_confirmation ... }`, but that's only moving the `Future` down the line, and the closure can't return one.

Reply:
Almost forgot: the `impl<T: Transport> Future` for `SendTransactionWithConfirmation<T>` has `type Item = TransactionReceipt`, meaning even if the above is done, we will still be getting a `TransactionReceipt` where we should be getting a `SendTransactionWithConfirmation<T>`.

Reply:
Most likely we need to change `PendingContract` to be able to work with aribtrary `Future` that resolves to `TransactionReceipt`:
```
PendingContract<
  T: Transport, 
  F: Future<Item=TransactionReceipt, Error=...> = SendTransactionWithConfirmation<T>,
> {

}
```
and then use the `and_then`combinator as you've mentioned.

Reply:
I was reluctant to change `PendingContract` without the go-ahead of someone more knowledgeable about the code base than me. If it sounds reasonable, then I will do so, and re-request a review.

Thank you for your attention :+1:

Reply:
I understand that the type signatures you wrote here are merely suggestive, but what does `F: Future<Item=TransactionReceipt, Error=...> = SendTransactionWithConfirmation<T>` mean? I have not come across this syntax before.

Reply:
It's a generic parameter, but with a default value. So that `PendingContract<T>` would still be backward compatible, as the `F` type would default to `SendTransactionWithConfirmation` if omitted.

Reply:
I split the two functions back again since the previous refactoring would not work with this change to `PendingContract`.

However, now I'm getting an error on [this line](https://github.com/tomusdrw/rust-web3/pull/252/files#diff-0dfcbf3acd5476076104fa446de24e33R177) I am not sure I understand:

```rust
F
mismatched types

expected type parameter, found struct `futures::AndThen`

note: expected type `Ft`
         found type `futures::AndThen<helpers::CallFuture<types::transaction::RawTransaction, <T as Transport>::Out>, confirm::SendTransactionWithConfirmation<T>, [closure@src\contract\deploy.rs:165:79: 172:10 eth:_, self:_]>`rustc(E0308)
deploy.rs(177, 13): expected type parameter, found struct `futures::AndThen`
```

I was under the impression `and_then` returned a struct, `AndThen`, that implemented `Future` so I am not sure why `PendingContract` would have issues with it.

Do you mind lending me a hand with this? I apologise for my unreasonable request.